File: software/DBSLMM.R
#! /usr/bin/env Rscript

########################################################################
# Deterministic Bayesian Sparse Linear Mixed Model (DBSLMM v1.0)       #
# Copyright (C) 2024  Sheng Yang and Xiang Zhou                        #
#                                                                      #
# This program is free software: you can redistribute it and/or modify #
# it under the terms of the GNU General Public License as published by #
# the Free Software Foundation, either version 3 of the License, or    #
# (at your option) any later version.                                  #
#                                                                      #
# This program is distributed in the hope that it will be useful,      #
# but WITHOUT ANY WARRANTY; without even the implied warranty of       #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the        #
# GNU General Public License for more details.                         #
#                                                                      #
# You should have received a copy of the GNU General Public License    #
# along with this program. If not, see <http://www.gnu.org/licenses/>. #
########################################################################

# Load packages
library(bigsnpr)
library(bigreadr)
library(dplyr)
library(plyr)
library(optparse)

# Set parameters
args_list <- list(
  ## Required parameters
  make_option("--summary", type = "character", default = NULL,
              help = "INPUT: the summary statistics (gemma output format)", metavar = "character"),
  make_option("--dbslmm", type = "character", default = NULL,
              help = "INPUT: the perfix of dbslmm software", metavar = "character"),
  make_option("--type", type = "character", default = "auto",
              help="INPUT: type of DBSLMM (default: default version)", metavar="character"),
  make_option("--model", type = "character", default = NULL,
              help = "INPUT: the chioce of LMM and DBSLMM", metavar = "character"),
  make_option("--reference", type = "character", default = NULL,
              help = "INPUT: the perfix of reference panel", metavar = "character"),
  make_option("--block", type = "character", default = NULL,
              help = "INPUT: the block information (Berisa and Pickrell 2015)", metavar = "character"),
  make_option("--N", type = "character", default = NULL,
              help="INPUT: sample size for summary statistics", metavar="character"),
  make_option("--outPath", type = "character", default = NULL,
              help="INPUT: the output path", metavar = "character"),
  ## Non-required parameters
  make_option("--validation_g", type = "character", default = NULL,
              help = "INPUT: the perfix of genotype of validation set", metavar = "character"),
  make_option("--validation_p", type = "character", default = NULL,
              help = "INPUT: the perfix of phenotype of validation set", metavar = "character"),
  make_option("--h2f", type = "character", default = NULL,
              help = "INPUT: the fold of heritability of trait", metavar = "character"),
  make_option("--mafMax", type = "character", default = "0.2",
              help = "INPUT: the maximium of the difference between reference panel and summary data (default:0.2)", 
              metavar = "character"),
  make_option("--thread", type = "character", default = "1",
              help = "INPUT: the number of threads (default: 1)", 
              metavar = "character")
)
opt_parser <- OptionParser(option_list=args_list)
opt <- parse_args(opt_parser)

# opt <- list(summary = "/public/home/biostat03/biosoft/DBSLMM/PGC.assoc.txt", 
#             dbslmm = "/public/home/biostat03/biosoft/DBSLMM/scr/dbslmm",
#             type = "auto", 
#             model = "LMM", 
#             block = "/public/home/biostat03/biosoft/DBSLMM/block_data/EAS/",
#             reference = "/public/home/biostat03/project/pgsfusionProject/ref_geno/", 
#             N = "4479,75725",
#             outPath = "/public/home/biostat03/biosoft/DBSLMM/tmp_output/", 
#             mafMax = "0.2", 
#             thread = "1")

# Check the software
if (!file.exists(opt$dbslmm)){
  cat(paste0("ERROR: ", opt$dbslmm, " does not exist! Please check!\n"))
  q()
}

# Check files
if (!file.exists(opt$block)){
  cat(paste0("ERROR: ", opt$block, " does not exist! Please check!\n"))
  q()
}
if (!file.exists(opt$summary)){
  cat(paste0("ERROR: ", opt$summary, " does not exist! Please check!\n"))
  q()
}
if (!file.exists(opt$outPath)){
  cat(paste0("ERROR: ", opt$outPath, " does not exist! Please check!\n"))
  q()
}
if (!file.exists(opt$reference)){
  cat(paste0("ERROR: ", opt$outPath, " does not exist! Please check!\n"))
  q()
}

# Check settings
if (opt$type == "tuning"){

  if (!file.exists(opt$validation_g) | !file.exists(opt$validation_p)){
    cat(paste0("ERROR: Validation set does not exist! Please check!\n"))
    q()
  }
  if(is.null(opt$h2f)){
    cat(paste0("ERROR: Fold does not set! Please check!\n"))
    q()
  }
}

# Set parameters
R2 <- 0.2
PVAL <- 1e-6
WIN <- 1000

# Estimate heritability for whole genome
map <- transmute(readRDS(paste0(opt$reference,"/map_hm3_plus.rds")),
                 chr = as.integer(chr), pos = pos, rsid, af_val = af_ref,
                 a0 = a0, a1 = a1, ldsc = ldsc)
sumstats <- fread2(opt$summary, header = T)
sumstats <- sumstats[order(sumstats[, 1], sumstats[, 3]), ]
sumstats_gemma <- sumstats
## Calculate effective sample size
N_str <- strsplit(opt$N, ",")[[1]] %>% as.numeric
if (length(N_str) == 1){
  
  n_eff <- N_str
} else {
  
  n_eff <- 4 / (1 / N_str[1] + 1 / N_str[2])
}
sumstats <- sumstats[, c(1, 3, 2, 6, 7, 8, 9, 10, 11)]
sumstats$n_eff <- n_eff
colnames(sumstats) <- c("chr", "pos", "rsid", "a0", "a1", "freq", "beta",
                        "beta_se", "pval", "n_eff")
## Coordinate summary statistics and map
df_beta <- as_tibble(snp_match(sumstats, map,
                               return_flip_and_rev = TRUE))
ldsc <- with(df_beta, snp_ldsc(ldsc, length(ldsc), chi2 = (beta / beta_se)^2,
                               sample_size = n_eff, blocks = NULL))
##
prefix_file <- strsplit(basename(opt$summary), "\\.")[[1]][1]

# DBSLMM-lmm
if (opt$model == "LMM"){
  
  lmm <- alply(c(1: 22), 1, function(CHR){
    
    sumstats_chr <- sumstats_gemma[sumstats_gemma$chr == CHR, ]
    chr_str <- paste0(opt$outPath, prefix_file, "_chr", CHR, ".assoc.txt")
    fwrite2(sumstats_chr, file = chr_str, col.names = F, sep = "\t")
    lmm_cmd <- paste0(opt$dbslmm,
                      " -s ",      chr_str,
                      " -r ",      paste0(opt$reference, "/merge"),
                      " -nsnp ",   length(df_beta),
                      " -n ",      as.integer(n_eff),
                      " -b ",      paste0(opt$block, "chr", CHR, ".bed"),
                      " -mafMax ", opt$mafMax,      
                      " -t ",      opt$thread)
    if (opt$type == "auto"){
      ## auto model
      r_lmm_cmd <- paste0(lmm_cmd, 
                          " -h ",      ldsc[[2]],
                          " -eff ",    opt$outPath, prefix_file, "_chr", CHR, ".dbslmm") %>% system()
    } else {
      ## tuning model
      h2_vec <- as.numeric(unlist(strsplit(opt$h2f, ",")))
      for (hh in h2_vec) {
        
        
        fwrite2(sumstats_chr, file = chr_str, col.names = F, sep = "\t")
        r_lmm_cmd <- paste0(lmm_cmd, 
                            " -h ",      ldsc[[2]] * hh,
                            " -eff ",    opt$outPath, prefix_file, "_chr", CHR,  "_h2f", hh, ".dbslmm") %>% system()
      }
    }
    
    ## remove temporary files
    system(paste0("rm -rf ", chr_str))
    system(paste0("rm -rf ", opt$outPath, prefix_file, "_chr", CHR, ".dbslmm.badsnps"))
    return(CHR)
  })
}

# DBSLMM model
if (opt$model == "DBSLMM"){
  
  ## C+T
  df_beta_sig <- df_beta[df_beta$pval < PVAL, ]
  if(!file.exists(paste0(opt$reference, "/merge.bk")))
    snp_readBed(paste0(opt$reference, "/merge.bed"))
  ref_bed <- snp_attach(paste0(opt$reference, "/merge.rds"))
  ref_sub_str <- paste0(opt$reference, "/merge_sub-", as.numeric(as.POSIXlt(Sys.time())))
  ref_sub_bed <- snp_attach(snp_subset(ref_bed, 
                                       ind.col = df_beta_sig$`_NUM_ID_`, 
                                       backingfile = ref_sub_str))
  ref_G <- snp_fastImputeSimple(ref_sub_bed$genotypes)
  lp_val <- -log10(df_beta_sig$pval)
  all_keep <- snp_grid_clumping(ref_G,
                                grid.thr.r2 = R2,
                                grid.base.size = WIN,
                                infos.chr = df_beta_sig$chr,
                                infos.pos = df_beta_sig$pos,
                                lpS = lp_val,
                                ncores = 1)
  system(paste0("rm -rf ", ref_sub_str))
  l_chr <- data.frame(chr = c(1: 22), 
                      l_b = c(1: 22) %in% names(all_keep), 
                      count = NA)
  l_chr$count[l_chr$l_b] <- c(1: sum(l_chr$l_b))
  ### Fit model
  dbslmm <- alply(c(1: 22), 1, function(CHR){
    
    dbslmm_cmd <- paste0(opt$dbslmm,
                         " -r ",      paste0(opt$reference, "/merge"),
                         " -nsnp ",   length(df_beta),
                         " -n ",      as.integer(n_eff),
                         " -b ",      paste0(opt$block, "chr", CHR, ".bed"),
                         " -mafMax ", opt$mafMax,      
                         " -t ",      opt$thread)
    
    if(CHR %in% names(all_keep)){
      
      l_rsid_chr <- df_beta_sig$rsid.ss[all_keep[[l_chr$count[l_chr$chr == CHR]]][[1]]]
      l_summ_chr <- sumstats_gemma[sumstats_gemma$rs %in% l_rsid_chr, ]
      l_chr_str <- paste0(opt$outPath, "l_", prefix_file, "_chr", CHR, ".assoc.txt")
      s_summ_chr <- sumstats_gemma %>%
        filter(chr == CHR & !sumstats_gemma$rs %in% l_rsid_chr)
      s_chr_str <- paste0(opt$outPath, "s_", prefix_file, "_chr", CHR, ".assoc.txt")
      fwrite2(l_summ_chr, file = l_chr_str, col.names = F, sep = "\t")
      fwrite2(s_summ_chr, file = s_chr_str, col.names = F, sep = "\t")
      if (opt$type == "auto"){
        ## auto model
        r_dbslmm_cmd <- paste0(dbslmm_cmd, 
                               " -l ",     l_chr_str, 
                               " -s ",     s_chr_str, 
                               " -h ",     ldsc[[2]],
                               " -eff ",   opt$outPath, prefix_file, "_chr", CHR, ".dbslmm") %>% system()
        
      } else {
        ## tuning model
        h2_vec <- as.numeric(unlist(strsplit(opt$h2f, ",")))
        for (hh in h2_vec) {
          
          r_dbslmm_cmd <- paste0(lmm_cmd, 
                                 " -l ",      l_chr_str, 
                                 " -s ",      s_chr_str, 
                                 " -h ",      ldsc[[2]] * hh,
                                 " -eff ",    opt$outPath, prefix_file, "_chr", CHR,  "_h2f", hh, ".dbslmm") %>% system()
        }
      }
      
      system(paste0("rm -rf ", l_chr_str))
    }else{
      
      sumstats_chr <- sumstats_gemma[sumstats_gemma$chr == CHR, ]
      s_chr_str <- paste0(opt$outPath, "s_", prefix_file, "_chr", CHR, ".assoc.txt")
      fwrite2(sumstats_chr, file = s_chr_str, col.names = F, sep = "\t")
      if (opt$type == "auto"){
        ## auto model
        r_dbslmm_cmd <- paste0(dbslmm_cmd, 
                               " -s ",     s_chr_str, 
                               " -h ",     ldsc[[2]],
                               " -eff ",   opt$outPath, prefix_file, "_chr", CHR, ".dbslmm") %>% system()
      } else {
        ## tuning model
        h2_vec <- as.numeric(unlist(strsplit(opt$h2f, ",")))
        for (hh in h2_vec) {
          
          r_dbslmm_cmd <- paste0(lmm_cmd, 
                                 " -s ",      s_chr_str, 
                                 " -h ",      ldsc[[2]] * hh,
                                 " -eff ",    opt$outPath, prefix_file, "_chr", CHR,  "_h2f", hh, ".dbslmm") %>% system()
          
        }
      }
    }
    
    ## remove temporary files
    system(paste0("rm -rf ", s_chr_str))
    system(paste0("rm -rf ", opt$outPath, prefix_file, "_chr", CHR, ".dbslmm.badsnps"))
    return(CHR)
  })
}

# # Format files
# paste0("cat ", opt$outPath, prefix_file, "_chr{1..22}.dbslmm.txt > ", 
#        opt$outPath, prefix_file, ".dbslmm") %>% system()
# for (chr in 1: 22)
#   paste0("rm -rf ", opt$outPath, prefix_file, "_chr", chr, ".dbslmm") %>% system()

File: scr/dbslmm.cpp
/*
Deterministic Bayesian Sparse Linear Mixed Model (DBSLMM)
Copyright (C) 2019  Sheng Yang and Xiang Zhou

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <vector>
#include <string>
#include <iostream>
#include <ctime>
#include <math.h>

#include "dtpr.hpp"
#include "dbslmmfit.hpp"
#include "dbslmm.hpp"

using namespace std;

DBSLMM::DBSLMM(void) :
	version("0.2"), date("07/25/2019"), year("2019")
{}

void DBSLMM::printHeader(void)
{
	cout << endl;
	cout << "*************************************************************"<< endl;
	cout << "  Deterministic Bayesian Sparse Linear Mixed Model (DBSLMM)  " << endl;
	cout << "  Version " << version << ", " << date << "                  " << endl;
	cout << "  Visit http://www.xzlab.org/software.html For Update        " << endl;
	cout << "  (C) " << year << " Sheng Yang, Xiang Zhou                  " << endl;
	cout << "  GNU General Public License                                 " << endl;
	cout << "  For Help, Type ./dbslmm -h                                 " << endl;
	cout << "*************************************************************" << endl;
	cout << endl;

	return;
}

void DBSLMM::printHelp(void) {
	cout << " FILE I/O RELATED OPTIONS" << endl;
	cout << " -s        [filename]  " << " specify input the summary data for the small effect SNPs." << endl;
	cout << " -l        [filename]  " << " specify input the summary data for the large effect SNPs." << endl;
	cout << " -r        [filename]  " << " specify input the bfile of reference data." << endl;
	cout << " -n        [num]       " << " specify input the sample size of the summary data." << endl;
	cout << " -mafMax   [num]       " << " specify input the maximium of the difference between reference panel and summary data." << endl;
	cout << " -nsnp     [filename]  " << " specify input the number of snp." << endl;
	cout << " -b        [num]       " << " specify input the block information." << endl;
	cout << " -h        [num]       " << " specify input the heritability." << endl;
	cout << " -t        [filename]  " << " specify input thread." << endl;
	cout << " -eff      [filename]  " << " specify output the estimate effect SNPs." << endl;
	return;
}

void DBSLMM::Assign(int argc, char ** argv, PARAM &cPar) {
	
	string str;
	for (int i = 0; i < argc; i++) {

		if (strcmp(argv[i], "--smallEff") == 0 || strcmp(argv[i], "-s") == 0) {

			if (argv[i + 1] == NULL || argv[i + 1][0] == '-') { continue; }
			++i;
			str.clear();
			str.assign(argv[i]);
			cPar.s = str;
		}
		else if (strcmp(argv[i], "--largeEff") == 0 || strcmp(argv[i], "-l") == 0) {

			if (argv[i + 1] == NULL || argv[i + 1][0] == '-') { continue; }
			++i;
			str.clear();
			str.assign(argv[i]);
			cPar.l = str;
		}
		else if (strcmp(argv[i], "--reference") == 0 || strcmp(argv[i], "-r") == 0) {

			if (argv[i + 1] == NULL || argv[i + 1][0] == '-') { continue; }
			++i;
			str.clear();
			str.assign(argv[i]);
			cPar.r = str;
		}
		else if (strcmp(argv[i], "--N") == 0 || strcmp(argv[i], "-n") == 0) {

			if (argv[i + 1] == NULL || argv[i + 1][0] == '-') { continue; }
			++i;
			str.clear();
			str.assign(argv[i]);
			cPar.n = atoi(str.c_str());
		}
		else if (strcmp(argv[i], "--mafMax") == 0 || strcmp(argv[i], "-mafMax") == 0) {

			if (argv[i + 1] == NULL || argv[i + 1][0] == '-') { continue; }
			++i;
			str.clear();
			str.assign(argv[i]);
			cPar.mafMax = atof(str.c_str());
		}
		else if (strcmp(argv[i], "--numSNP") == 0 || strcmp(argv[i], "-nsnp") == 0) {

			if (argv[i + 1] == NULL || argv[i + 1][0] == '-') { continue; }
			++i;
			str.clear();
			str.assign(argv[i]);
			cPar.nsnp = atoi(str.c_str());
		}
		else if (strcmp(argv[i], "--block") == 0 || strcmp(argv[i], "-b") == 0) {

			if (argv[i + 1] == NULL || argv[i + 1][0] == '-') { continue; }
			++i;
			str.clear();
			str.assign(argv[i]);
			cPar.b = str;
		}
		else if (strcmp(argv[i], "--Heritability") == 0 || strcmp(argv[i], "-h") == 0) {

			if (argv[i + 1] == NULL || argv[i + 1][0] == '-') { continue; }
			++i;
			str.clear();
			str.assign(argv[i]);
			cPar.h = atof(str.c_str());
		}
		else if (strcmp(argv[i], "--Thread") == 0 || strcmp(argv[i], "-t") == 0) {

			if (argv[i + 1] == NULL || argv[i + 1][0] == '-') { continue; }
			++i;
			str.clear();
			str.assign(argv[i]);
			cPar.t = atoi(str.c_str());
		}
		else if (strcmp(argv[i], "--EFF") == 0 || strcmp(argv[i], "-eff") == 0) {

			if (argv[i + 1] == NULL || argv[i + 1][0] == '-') { continue; }
			++i;
			str.clear();
			str.assign(argv[i]);
			cPar.eff = str;
		}
	}
	return;
}

void DBSLMM::BatchRun(PARAM &cPar) {

	SNPPROC cSP;
	IO cIO;
	DBSLMMFIT cDBSF;

	// input check
	cout << "Options: " << endl;
	cout << "-s:      " << cPar.s << endl;
	cout << "-l:      " << cPar.l << endl;
	cout << "-r:      " << cPar.r << endl;
	cout << "-nsnp:   " << cPar.nsnp << endl;
	cout << "-n:      " << cPar.n << endl;
	cout << "-mafMax: " << cPar.mafMax << endl;
	cout << "-b:      " << cPar.b << endl;
	cout << "-h:      " << cPar.h << endl;
	cout << "-t:      " << cPar.t << endl;
	cout << "-eff:    " << cPar.eff << endl;
	
	// check files
	string ref_fam_str = cPar.r + ".fam";
	ifstream seffStream(cPar.s.c_str()), leffstream(cPar.l.c_str()), reffstream(ref_fam_str.c_str()), beffstream(cPar.b.c_str());
	if (cPar.s.size() == 0) {
		cerr << "ERROR: -s is no parameter!" << endl;
		exit(1);
	}
	if (cPar.l.size() == 0) {
		cerr << "ERROR: -l is no parameter!" << endl;
		exit(1);
	}
	if (!beffstream) {
		cerr << "ERROR: " << cPar.b << " dose not exist!" << endl;
		exit(1);
	}
	if (!seffStream) {
		cerr << "ERROR: " << cPar.s << " dose not exist!" << endl;
		exit(1);
	}
	if (!reffstream || cPar.l.size() == 0) {
		cerr << "ERROR: " << cPar.r << " dose not exist!" << endl;
		exit(1);
	}
	if (cPar.b.size() == 0) {
		cerr << "ERROR: -b is no parameter!" << endl;
		exit(1);
	}
	if (cPar.r.size() == 0) {
		cerr << "ERROR: " << cPar.r << " dose not exist!" << endl;
		exit(1);
	}
	if (cPar.h > 1 || cPar.h < 0){
		cerr << "ERROR: -h is not correct (0, 1)!" << endl;
		exit(1);
	}
	if (cPar.t > 100 || cPar.t < 1){
		cerr << "ERROR: -t is not correct (1, 100)!" << endl;
		exit(1);
	}
	
	// get sample size of reference panel 
	char separate[] = "\t";
	cout << "Reading PLINK FAM file from [" << cPar.r << ".fam]" << endl;
	int n_ref = cIO.getRow(ref_fam_str);
	cout << n_ref << " individuals to be included from reference FAM file." << endl;
	
	// get SNP of reference panel
	cout << "Reading PLINK BIM file from [" << cPar.r << ".bim]" << endl;
	map <string, ALLELE> ref_bim;
	bool constr = true; 
	if (abs(cPar.mafMax-1.0) < 1e-10){
		constr = false; 
	}
	cIO.readBim(n_ref, cPar.r, separate, ref_bim, constr);
	int num_snp_ref = ref_bim.size();
	cout << num_snp_ref << " SNPs to be included from reference BIM file." << endl;

	// input block file
	vector <BLOCK> block_dat; 
	cIO.readBlock(cPar.b, separate, block_dat);
 
	// input small effect summary data
	cout << "Reading summary data of small effect SNPs from [" << cPar.s << "]" << endl;
	vector <SUMM> summ_s;
	int n_s = cIO.readSumm(cPar.s, separate, summ_s);
	vector <POS> inter_s;
	bool badsnp_s[n_s] = {false}; 
	cSP.matchRef(summ_s, ref_bim, inter_s, cPar.mafMax, badsnp_s);
	cout << "After filtering, " << inter_s.size() << " small effect SNPs are selected." << endl;
	vector <INFO> info_s; 
	int num_block_s = cSP.addBlock(inter_s, block_dat, info_s); 

	// input large effect summary data
	cout << "Reading summary data of large effect SNPs from [" << cPar.l << "]" << endl;
	vector <SUMM> summ_l;
	int n_l = cIO.readSumm(cPar.l, separate, summ_l);
	vector <POS> inter_l;
	bool badsnp_l[n_l] = {false}; 
	cSP.matchRef(summ_l, ref_bim, inter_l, cPar.mafMax, badsnp_l);
	vector <INFO> info_l; 
	if (inter_l.size() != 0){
		int num_block_l = cSP.addBlock(inter_l, block_dat, info_l); 
		cout << "After filtering, " << inter_l.size() << " large effect SNPs are selected." << endl;
	} else {
		cout << "After filtering, no large effect SNP is selected." << endl;
	}
	
	// output badsnps
	string badsnps_str = cPar.eff + ".badsnps"; 
	ofstream badsnpsFout(badsnps_str.c_str());
	for (size_t i = 0; i < summ_l.size(); ++i) {
		if (badsnp_l[i] == false){
			badsnpsFout << summ_l[i].snp << " " << 1 << endl;
		}
	}
	clearVector(summ_l);
	for (size_t i = 0; i < summ_s.size(); ++i) {
		if (badsnp_s[i] == false)
			badsnpsFout << summ_s[i].snp << " " << 0 << endl;
	}
	clearVector(summ_s);
	
	// output stream
	string eff_str = cPar.eff + ".txt"; 
	ofstream effFout(eff_str.c_str());
	if (inter_l.size() != 0){
		// fit model
		vector <EFF> eff_s, eff_l; 
		vector<int> idv(n_ref);
		for (int i = 0; i < n_ref; i++) idv[i] = 1; 
		string bed_str = cPar.r + ".bed";
		double t_fitting = cIO.getWalltime();
		double sigma_s = cPar.h / (double)cPar.nsnp;
		cout << "Fitting model..." << endl;
		cDBSF.est(n_ref, cPar.n, sigma_s, num_block_s, idv, bed_str, info_s, info_l, cPar.t, eff_s, eff_l); 
		double time_fitting = cIO.getWalltime() - t_fitting;
		cout << "Fitting time: " << time_fitting << " seconds." << endl;

		// output effect 
		for (size_t i = 0; i < eff_l.size(); ++i) {
			double beta_l_noscl = eff_l[i].beta / sqrt(2 * eff_l[i].maf * (1-eff_l[i].maf));
			if (eff_l[i].snp != "rs" && isinf(beta_l_noscl) == false)
				effFout << eff_l[i].snp << " " << eff_l[i].a1 << " " << eff_l[i].beta << " " << beta_l_noscl << " " << 1 << endl; 
		}
		
		for (size_t i = 0; i < eff_s.size(); ++i) { 
			double beta_s_noscl = eff_s[i].beta / sqrt(2 * eff_s[i].maf * (1-eff_s[i].maf));
			if(eff_s[i].snp.size() != 0 && isinf(beta_s_noscl) == false)
				effFout << eff_s[i].snp << " " << eff_s[i].a1 << " " << eff_s[i].beta << " " << beta_s_noscl << " " << 0 << endl; 
		}
		effFout.close();
	} else {
		// fit model
		vector <EFF> eff_s; 
		vector<int> idv(n_ref);
		for (int i = 0; i < n_ref; i++) idv[i] = 1; 
		string bed_str = cPar.r + ".bed";
		double t_fitting = cIO.getWalltime();
		double sigma_s = cPar.h / (double)cPar.nsnp;
		cout << "Fitting model..." << endl;
		cDBSF.est(n_ref, cPar.n, sigma_s, num_block_s, idv, bed_str, info_s, cPar.t, eff_s); 
		double time_fitting = cIO.getWalltime() - t_fitting;
		cout << "Fitting time: " << time_fitting << " seconds." << endl;

		// output effect 
		for (size_t i = 0; i < eff_s.size(); ++i) { 
			double beta_s_noscl = eff_s[i].beta / sqrt(2 * eff_s[i].maf * (1-eff_s[i].maf));
			if(eff_s[i].snp.size() != 0 && isinf(beta_s_noscl) == false)
				effFout << eff_s[i].snp << " " << eff_s[i].a1 << " " << eff_s[i].beta << " " << beta_s_noscl << " " << 0 << endl; 
		}
	}
	return;
}
